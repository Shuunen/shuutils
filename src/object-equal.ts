import { objectSerialize } from './object-serializer'

/**
 * Check if two objects are deeply equal
 * @param objectA the base object to compare
 * @param objectB the object to compare with
 * @param willSortKeys if true, the order of keys will be sorted alpha before comparison
 * @returns true if the two objects are deeply equal
 */
export function objectEqual (objectA: Readonly<Record<string, unknown>>, objectB: Readonly<Record<string, unknown>>, willSortKeys = false) {
  return objectSerialize(objectA, willSortKeys) === objectSerialize(objectB, willSortKeys)
}
