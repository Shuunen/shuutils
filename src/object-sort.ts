// Based on : https://github.com/bevry/sortobject/blob/master/source/index.ts
// Copyright Â© Benjamin Lupton
// Artistic License 2.0

/* eslint-disable @typescript-eslint/consistent-type-assertions */

/** A typical comparator for sorting object keys */
type KeyComparator = (a: string, b: string) => number

/** An indexable object */
type IndexedObject = Record<string, unknown>

/**
 * Returns a copy of the passed array, with all nested objects within it sorted deeply by their keys, without mangling any nested arrays.
 * @param subject the unsorted array
 * @param comparator an optional comparator for sorting keys of objects
 * @returns the new sorted array
 */
export function arraySort<ArrayType extends unknown[]> (subject: ArrayType, comparator?: KeyComparator) {
  const result = []
  for (let value of subject) {
    if (value !== null && value !== undefined)
      if (Array.isArray(value)) value = arraySort(value, comparator) // eslint-disable-next-line @typescript-eslint/no-use-before-define
      else if (typeof value === 'object') value = objectSort(value as IndexedObject, comparator)
    result.push(value)
  }
  return result as ArrayType
}

/**
 * Returns a copy of the passed object, with all nested objects within it sorted deeply by their keys, without mangling any nested arrays inside of it.
 * @param subject the unsorted object
 * @param comparator an optional comparator for sorting keys of objects
 * @returns the new sorted object
 */
export function objectSort<ObjectType extends IndexedObject> (subject: ObjectType, comparator?: KeyComparator) {
  const result: IndexedObject = {} as ObjectType
  const sortedKeys = Object.keys(subject).sort(comparator)
  for (const key of sortedKeys) {
    let value = subject[key]
    if (value !== null && value !== undefined)
      if (Array.isArray(value)) value = arraySort(value, comparator)
      else if (typeof value === 'object') value = objectSort(value as IndexedObject, comparator)
    result[key] = value
  }
  return result as ObjectType
}
