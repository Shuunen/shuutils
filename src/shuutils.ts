export * from './analytics'
export * from './array-insert-after'
export * from './array-pick-one'
export * from './array-remove-value'
export * from './array-shuffle'
export * from './array-unique'
export * from './browser-clipboard'
export * from './browser-fill-like-human'
export * from './browser-inject-styles'
export * from './browser-page-change'
export * from './browser-routing'
export * from './browser-scout'
export * from './browser-styles'
export * from './browser-toast'
export * from './colors'
export * from './constants'
export * from './date-relative'
export * from './dates'
export * from './debounce'
export * from './dom'
export * from './environment'
export * from './events'
export * from './expect-type'
export * from './flags'
export * from './functions'
export * from './logger'
export * from './memoize'
export * from './number-ranged-score'
export * from './number-round'
export * from './object-cap-nesting'
export * from './object-clone'
export * from './object-equal'
export * from './object-flatten'
export * from './object-serializer'
export * from './object-sort'
export * from './objects'
export * from './random'
export * from './result'
export * from './state'
export * from './storage'
export * from './string-case'
export * from './string-remove-accents'
export * from './strings'
export * from './system'
export * from './testing-align-for-snap'
export * from './throttle'
export * from './timestamps'
export * from './translate'
export type * from './types'
