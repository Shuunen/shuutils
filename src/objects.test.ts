/* eslint-disable unicorn/no-null */
/* eslint-disable jsdoc/require-jsdoc */
/* eslint-disable @typescript-eslint/naming-convention */
import { expect, it } from 'bun:test'
import { objectEqual } from './object-equal'
import { objectSort } from './object-sort'
import { access, byProperty, clone, flatten, genClass, isRecord, objectSum } from './shuutils'

const person = {
  age: 21,
  details: { dateOfBirth: new Date('2001-12-22'), favoriteFood: 'sushi' },
  name: 'John',
  nameRegex: /^jo/iu,
  nameValid: true,
  nameValidator: (input: string) => input.length > 3,
}
const personClone = clone(person)
const personCloneModified = clone(person)
personCloneModified.age = 42
it('clone A record', () => {
  expect(personCloneModified.age).toBe(42)
})
it('clone B record does not affect the original one', () => {
  expect(person.age).toBe(21)
})

const persons = ['John', 'Fanny']
const personsCopy = clone(persons)
personsCopy[1] = 'Bob'
it('clone C array', () => {
  expect(personsCopy[1]).toBe('Bob')
})
it('clone D array does not affect the original one', () => {
  expect(persons[1]).toBe('Fanny')
})
it('clone E original and clone are have the same objectSum', () => {
  expect(objectSum(person)).toBe(objectSum(personClone))
})
it('clone F original and clone are equals via objectEqual', () => {
  expect(objectEqual(person, personClone)).toBe(true)
})
it('clone G two clones are equals via objectEqual', () => {
  expect(objectEqual(clone(person), clone(person))).toBe(true)
})
it('clone H cloned values keep their type', () => {
  const cloned = clone(person)
  expect(typeof cloned.nameValidator).toBe('function')
  expect(typeof cloned.nameRegex).toBe('object')
  expect(cloned.nameRegex instanceof RegExp).toBe(true)
  expect(typeof cloned.details.dateOfBirth).toBe('object')
  expect(cloned.details.dateOfBirth instanceof Date).toBe(true)
})
it('clone I debug', () => {
  expect({ person, personCopy: personCloneModified }).toMatchSnapshot()
})

it('access A nested existing property access', () => {
  expect(access(person, 'details.favoriteFood')).toBe('sushi')
})
it('access B nested non-existing property access', () => {
  expect(typeof access(person, 'details.favoriteDrink')).toBe('undefined')
})
it('access C non-nested property', () => {
  expect(access({ name: 'John Cena' }, 'name')).toBe('John Cena')
})
it('access D non-existing non-nested property', () => {
  expect(typeof access({ name: 'John Cena' }, 'age')).toBe('undefined')
})
it('access E non-nested property after an undefined property', () => {
  expect(access({ age: undefined, name: 'John Cena' }, 'name')).toBe('John Cena')
})

it('flatten A object', () => {
  expect(flatten(person)).toMatchSnapshot()
})
it('flatten B object with a custom root path', () => {
  expect(flatten(person, 'person')).toMatchSnapshot()
})
it('flatten C object containing an array', () => {
  expect(flatten({ collection: ['pikachu', 'drake'], name: 'John' })).toMatchSnapshot()
})

const users = [
  { age: 21, name: 'John', pic: 'wow.png' },
  { age: 42, name: 'Albert' },
  { age: 22, name: 'Sam' },
  { age: 11, name: 'Birgit' },
]
it('sort byProperty A without order does not sort', () => {
  expect(users.sort(byProperty('name'))[0]?.name).toBe('John')
})
it('sort byProperty B with asc order does sort', () => {
  expect(users.sort(byProperty('name', 'asc'))[0]?.name).toBe('Albert')
})
it('sort byProperty C with desc order does sort', () => {
  expect(users.sort(byProperty('name', 'desc'))[0]?.name).toBe('Sam')
})
it('sort byProperty D even if some does not have it', () => {
  expect(users.sort(byProperty('pic', 'asc'))[0]?.pic).toBe('wow.png')
})

const object3 = { notFun: false, 'pretty-good': true, size: 'large', superFun: true }
it('class generator A undefined', () => {
  // eslint-disable-next-line unicorn/no-useless-undefined
  expect(genClass(undefined)).toBe('')
})
it('class generator B null', () => {
  expect(genClass(null)).toBe('')
})
it('class generator C empty string', () => {
  expect(genClass('')).toBe('')
})
it('class generator D string', () => {
  expect(genClass('one 2   three')).toBe('one 2 three')
})
it('class generator E string with cls', () => {
  expect(genClass('one 2   three', [], ['enabled', 2])).toBe('one 2 three enabled')
})
it('class generator F empty', () => {
  expect(genClass({})).toBe('')
})
it('class generator G object', () => {
  expect(genClass(object3)).toMatchInlineSnapshot('"pretty-good size-large superFun"')
})
it('class generator H object & specific keys', () => {
  expect(genClass(object3, ['pretty-good'])).toBe('pretty-good')
})
it('class generator I object & specific keys & custom class', () => {
  expect(genClass(object3, ['pretty-good'], ['nice ok'])).toBe('nice ok pretty-good')
})
it('class generator J object containing array', () => {
  expect(genClass({ status: ['visible', 'enabled'] })).toBe('has-status')
})
it('class generator K object containing empty array', () => {
  expect(genClass({ status: [] })).toBe('')
})
it('class generator L array', () => {
  expect(genClass(['one one ', 2, null, 'one', undefined, '   three '])).toBe('one 2 three')
})

it('isRecord A on null', () => {
  expect(isRecord(null)).toBe(false)
})
it('isRecord B on an array', () => {
  expect(isRecord([1, 2])).toBe(false)
})
it('isRecord C on an empty array', () => {
  expect(isRecord([])).toBe(false)
})
it('isRecord D on an empty string', () => {
  expect(isRecord('')).toBe(false)
})
it('isRecord E on a number', () => {
  expect(isRecord(-1)).toBe(false)
})
it('isRecord F on an empty record', () => {
  expect(isRecord({})).toBe(true)
})
it('isRecord G on a record', () => {
  expect(isRecord({ name: 'John' })).toBe(true)
})

const object4 = { fool: { bar: 'foo', regex: /^ho\d+$/iu }, keyA: 1, keyB: 2, keyC: 3 }
const object4ButAnotherReference = { fool: { bar: 'foo', regex: /^ho\d+$/iu }, keyA: 1, keyB: 2, keyC: 3 }
const object4ButDeepRegexDifferent = { fool: { bar: 'foo', regex: /^oh\d+$/iu }, keyA: 1, keyB: 2, keyC: 3 }
it('objectSum A on empty object', () => {
  expect(objectSum({})).toMatchInlineSnapshot('2745614147')
})
it('objectSum B is the same on two equally empty objects', () => {
  expect(objectSum({ name: 'john' }) === objectSum(clone({ name: 'john' }))).toBe(true)
})
it('objectSum C on object with numbers', () => {
  expect(objectSum({ keyA: 1, keyB: 2, keyC: 3 })).toMatchInlineSnapshot('2822221177')
})
it('objectSum D on object with a slightly different number', () => {
  expect(objectSum({ keyA: 1, keyB: 2, keyC: 4 })).toMatchInlineSnapshot('3883285438')
})
it('objectSum E on object with a slightly different key', () => {
  expect(objectSum({ keyA: 1, keyC: 3, keyD: 2 })).toMatchInlineSnapshot('1603108510')
})
it('objectSum F on a large object', () => {
  expect(
    objectSum({
      abyss: 1,
      backInTime: 'was a good movie',
      clearlyHugeObjectThere: 33_514_149_687,
      details: {},
      object3,
      propE: 5,
      propF: 6,
      propG: 7,
      propH: 8,
      propI: 9,
      propJ: 10,
      users,
    }),
  ).toMatchInlineSnapshot('3077461858')
})

it('objectSum G is the same on two equals objects', () => {
  expect(objectSum(object3) === objectSum(clone(object3))).toBe(true)
})

it('objectSum H is the same on two equals objects with nested objects', () => {
  expect(objectSum(object4) === objectSum(object4ButAnotherReference)).toBe(true)
})

it('objectSum I is different with a deep slight modification', () => {
  expect(objectSum(object4) !== objectSum(object4ButDeepRegexDifferent)).toBe(true)
})

it('objectSum J changes depending on key order', () => {
  // eslint-disable-next-line perfectionist/sort-objects
  expect(objectSum({ keyA: 1, keyB: 2, keyC: 3 }) !== objectSum({ keyC: 3, keyA: 1, keyB: 2 })).toBe(true)
})

it('objectSum K same when key sorted', () => {
  // eslint-disable-next-line perfectionist/sort-objects
  expect(objectSum({ keyA: 1, keyB: 2, keyC: 3 }, true) === objectSum({ keyC: 3, keyA: 1, keyB: 2 }, true)).toBe(true)
})

it('objectEqual A true for empty objects', () => {
  expect(objectEqual({}, {})).toBe(true)
})

it('objectEqual B true for objects with same keys', () => {
  expect(objectEqual({ keyA: 1, keyB: 2, keyC: 3 }, { keyA: 1, keyB: 2, keyC: 3 })).toBe(true)
})

it('objectEqual C false for objects with different keys', () => {
  expect(objectEqual({ keyA: 1, keyB: 2, keyC: 3 }, { keyA: 1, keyB: 2, keyD: 3 })).toBe(false)
})
it('objectEqual D false for objects with different values', () => {
  expect(objectEqual({ keyA: 1, keyB: 2, keyC: 3 }, { keyA: 1, keyB: 2, keyC: 4 })).toBe(false)
})
it('objectEqual E true with objects containing same literals', () => {
  expect(objectEqual({ bar: 2, fool: 1 }, { bar: 2, fool: 1 })).toBe(true)
})
it('objectEqual F true for deeply nested objects', () => {
  expect(objectEqual({ fool: { bar: 'foo', regex: /^ho\d+$/iu } }, { fool: { bar: 'foo', regex: /^ho\d+$/iu } })).toBe(true)
})
it('objectEqual H false with objects containing different literals', () => {
  expect(objectEqual({ bar: 1, fool: 1 }, { bar: 2, fool: 1 })).toBe(false)
})
it('objectEqual I true with same objects but different references', () => {
  expect(objectEqual(object4, object4ButAnotherReference)).toBe(true)
})
it('objectEqual J false with same objects but different key order', () => {
  // eslint-disable-next-line perfectionist/sort-objects
  expect(objectEqual({ keyA: 1, keyB: 2, keyC: 3 }, { keyC: 3, keyA: 1, keyB: 2 })).toBe(false)
})
it('objectEqual K true with same objects, different key order but key sort is active', () => {
  // eslint-disable-next-line perfectionist/sort-objects
  expect(objectEqual({ keyA: 1, keyB: 2, keyC: 3 }, { keyC: 3, keyA: 1, keyB: 2 }, true)).toBe(true)
})

it('objectSort A sort a simple object', () => {
  const object = { keyC: 3, keyA: 1, keyB: 2 } // eslint-disable-line perfectionist/sort-objects
  const sorted = objectSort(object)
  // we cannot use toStrictEqual because toStrictEqual does not check the order of the keys, objectEqual does
  expect(objectEqual(sorted, { keyA: 1, keyB: 2, keyC: 3 })).toBe(true)
})

it('objectSort B sort a 2 level object', () => {
  const object = { keyC: 3, keyA: 1, alpaca: { c3: 33, c1: 11, c2: 22 }, keyB: 2 } // eslint-disable-line perfectionist/sort-objects
  const sorted = objectSort(object)
  expect(objectEqual(sorted, { alpaca: { c1: 11, c2: 22, c3: 33 }, keyA: 1, keyB: 2, keyC: 3 })).toBe(true)
})

it('objectSort C sort a complex object with null, undefined, etc', () => {
  /* eslint-disable perfectionist/sort-objects, unicorn/no-null */
  const object = {
    keyC: 3,
    keyA: undefined,
    alpaca: { c3: 33, c1: 11, c4: undefined, c2: 22 },
    names: ['john', null, [{ c3: 3, b2: 2 }], 'eddy'],
    keyB: 2,
    zebra: { z3: 33, z1: 11, z2: null },
  }
  const sorted = objectSort(object)
  expect(sorted).toMatchSnapshot()
})
