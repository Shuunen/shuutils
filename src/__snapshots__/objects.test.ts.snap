// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`clone I debug 1`] = `
{
  "person": {
    "age": 21,
    "details": {
      "dateOfBirth": 2001-12-22T00:00:00.000Z,
      "favoriteFood": "sushi",
    },
    "name": "John",
    "nameRegex": /^jo/iu,
    "nameValid": true,
    "nameValidator": [Function],
  },
  "personCopy": {
    "age": 42,
    "details": {
      "dateOfBirth": 2001-12-22T00:00:00.000Z,
      "favoriteFood": "sushi",
    },
    "name": "John",
    "nameRegex": /^jo/iu,
    "nameValid": true,
    "nameValidator": [Function],
  },
}
`;

exports[`flatten A object 1`] = `
{
  "age": 21,
  "details.dateOfBirth": 2001-12-22T00:00:00.000Z,
  "details.favoriteFood": "sushi",
  "name": "John",
  "nameRegex": /^jo/iu,
  "nameValid": true,
  "nameValidator": [Function],
}
`;

exports[`flatten B object with a custom root path 1`] = `
{
  "person.age": 21,
  "person.details.dateOfBirth": 2001-12-22T00:00:00.000Z,
  "person.details.favoriteFood": "sushi",
  "person.name": "John",
  "person.nameRegex": /^jo/iu,
  "person.nameValid": true,
  "person.nameValidator": [Function],
}
`;

exports[`flatten C object containing an array 1`] = `
{
  "collection[0]": "pikachu",
  "collection[1]": "drake",
  "name": "John",
}
`;

exports[`objectSort C sort a complex object with null, undefined, etc 1`] = `
{
  "alpaca": {
    "c1": 11,
    "c2": 22,
    "c3": 33,
    "c4": undefined,
  },
  "keyA": undefined,
  "keyB": 2,
  "keyC": 3,
  "names": [
    "john",
    null,
    [
      {
        "b2": 2,
        "c3": 3,
      },
    ],
    "eddy",
  ],
  "zebra": {
    "z1": 11,
    "z2": null,
    "z3": 33,
  },
}
`;
